### ISSUE TITLE
TASK: Potential fixup/followups for the allocator changes

### ISSUE THREAD

**Issue by seberg** [numpy/numpy#20193]
These is a list of potential followups for gh-17582

* [x] Release note is missing? PR #20194 
* [ ] I would like a bit "hand holding" for the arr->flags |= NPY_OWNDATA user. I.e. tell users in some place, (maybe the flag documentation?) linked from the warning: Please use use a PyCapsule as the array base (this is simple, but you have to know about PyCapsule). (I could look into that myself.)
* [ ] `PyCapsule_GetPointer` technically can set an error and may need the GIL? I am not worried enough about this to delay, but the functions look like they are meant to be callable without the GIL (I doubt we use it, and I doubt it really can error/need the GIL, but maybe we can make it a bit nicer anyway).
* [ ] The void functions use (normally) small allocation: I think it may make more sense to just use the Python allocators rather than a (possibly slow) aligned allocator.  (This is `VOID_compare`)
* [x] We should double check that we are OK without an explicit versioning mechanism (and maybe related with users assigning directly assigning the handler).

Anyone else have anything that might make sense to look into?

**Comment by jakirkham** [comment]
cc @pentschev @madsbk @quasiben (for awareness)

**Comment by eliaskoromilas** [comment]
> Something similar to what @eliaskoromilas did with his [numpy-allocator](https://github.com/inaccel/numpy-allocator) would do:
> 
> ```python
> >>> import numpy as np
> >>> 
> >>> # let downstream libraries/users worry about how to prepare the necessary interface objects
> >>> my_allocator = np.create_allocator(
> >>>     malloc=...,
> >>>     calloc=...,
> >>>     realloc=...,
> >>>     free=...)
> >>> with my_allocator:  # change the allocator locally
> ...     a = np.array([1, 2, 3])
> >>> 
> >>> # or, change the allocator globally
> >>> curr_allocator = np.get_allocator()
> >>> np.set_allocator(my_allocator)
> >>> b = np.array([4, 5, 6])
> ```

My [NumPy Allocator](https://github.com/inaccel/numpy-allocator) API actually fulfills all these requirements through:

### a) The allocator metaclass (`numpy_allocator.type`)

Let me first explain how metaclasses work (for those not familiar with the term). Classes in Python, by default, use `object` as their base and `type` as their metaclass, which means that the following (3) are equivalent:

```python
class <name>:
    pass

# or

class <name>(metaclass=type):
    pass

# or

<name> = type('<name>', (), dict())
```
Custom metaclasses can be used to provide static attributes to a class. In NumPy Allocator's case, `numpy_allocator.type` is responsible for initializing the handler capsule (based on the provided allocator funcs), but also for the context management functionality. It makes describing custom allocators as easy as:

```python
class <name>(metaclass=numpy_allocator.type):
    _calloc_ = <calloc>
    _free_ = <free>
    _malloc_ = <malloc>
    _realloc_ = <realloc>

# or

<name> = numpy_allocator.type(<name>, (), dict(_calloc_=<calloc>, _free_=<free>, _malloc_=<malloc>, _realloc_=<realloc>))
```
Of course you are free to wrap this with a function:

```python
def create_allocator(name, malloc=None, calloc=None, realloc=None, free=None):
    return numpy_allocator.type(name, (), dict(_calloc_=calloc, _free_=free, _malloc_=malloc, _realloc_=realloc))
```

It's important to note here that `ctypes` function pointers are used to define the `_{c/m/re]alloc_/_free_` functions (independently), which means that both Python callback functions and DLL symbols are allowed.

### b) The low-level handler API (`numpy_allocator.{g|s}et_handler`)

If you don't care about context management and the high-level interface that the metaclass offers, you can still use the low-level handler API to switch between allocators. If you already have a capsule containing the NumPy memory handler, you can use only that part of the library to get, set or reset the context-local handler, but also to extract the handler from a specific array.

```python
# let's assume that somehow we have got a valid "mem_handler" capsule (e.g. std_handler)

# std_handler could be also created like this
# class std_allocator(metaclass=numpy_allocator.type):
#     pass

# std_handler = std_allocator.handler()

#################################

numpy_allocator.set_handler(std_handler)

test = numpy.ndarray(())

print(numpy.core.multiarray.get_handler_name())  # prints: std_allocator

numpy_allocator.set_handler(None)

print(numpy.core.multiarray.get_handler_name())  # prints: default_allocator

numpy_allocator.set_handler(numpy_allocator.get_handler(test))

print(numpy.core.multiarray.get_handler_name())  # prints: std_allocator
```

**Comment by eliaskoromilas** [comment]
> ... about hooking allocators into Python, there are a few use cases that stick out to me:
> 
> * Allocators with specific alignment (some low-level function expect very specific alignment)

This is an [example](https://github.com/inaccel/numpy-allocator/blob/master/test/aligned_allocator.py#L23-L44) of how a page-aligned allocator can be written in Python, utilizing the NumPy Allocator API.

> * Shared memory (useful for working with large memory allocations in a process parallel friendly way)

This is an [example](https://github.com/inaccel/coral-api/blob/master/src/main/python/inaccel/coral/_shm.py) of how we (InAccel) use this API to integrate our shared memory architecture with NumPy. As you will notice, in this case we just open `libcoral-api` DLL and pass on the desired symbols to the respective `inaccel_allocator` class attributes.

**Comment by seberg** [comment]
@eliaskoromilas just wondering if you have a thought on how/whether we should have some versioning (e.g. a version number stored in the struct)?

**Comment by jakirkham** [comment]
Maybe there could be a method for getting that version? Agree having a versioned API is important (this API may well change in the future)

**Comment by seberg** [comment]
Hmm, means we need to do `if/else` chains.  Although, I guess you could also get the capsule name and do a `strncmp` only on the first part (up to the version), then check the version.
My main thought was that it might be nice if `alloc` and `free` don't bloat for this, but I suppose that is possible.

**Comment by leofang** [comment]
IIRC @seberg isn't a fan of using the capsule name to contain operational information 😄 But I can live with that. 

One nitpick on @eliaskoromilas's example is that `PyDataMemAllocator2` should be made ABI compatible with `PyDataMemAllocator` by adding `memcpy` to the end of the struct. Extra safety should be always preferred.

**Comment by seberg** [comment]
Yeah, but I can be convinced.  I don't like that `if/else` chain, but now I realize that can probably be somewhat avoided.  Not a huge fan, but so long version is the only thing we safe it is probably fine.
And I don't really expect more versions than maybe `2` and `3` ;).

And I suppose... we could even "deprecate" versions effectively, but making a very quick check in the `__dealloc__`, which is guaranteed to get called.

**Comment by eliaskoromilas** [comment]
> IIRC @seberg isn't a fan of using the capsule name to contain operational information smile But I can live with that.

I'm not a huge fan either. In the capsule name solution there is no compatibility between the different handler structs. In other words, we don't have "versions" but "identifiers".

> One nitpick on @eliaskoromilas's example is that `PyDataMemAllocator2` should be made ABI compatible with `PyDataMemAllocator` by adding `memcpy` to the end of the struct. Extra safety should be always preferred.

Intentionally I messed with the existing fields to note this functionality. If we choose to go with this solution, "mem_handler2" capsules won't be supported by NumPy versions that only know how to handle "mem_handler" capsules, etc. User libraries need to make sure that their allocator is compatible with the installed ΝumPy version.

> EDIT: And yes, I assume any newer version is ABI compatible with all older ones.

There is an alternate solution, of course, that focuses exactly on that. In this solution:

* we (MUST) keep the capsule name fixed ("mem_handler")
* we add a version field before/after the name field in the handler struct (needs to be done immediately, before 1.22.0) 
* newer NumPy versions are allowed only to **append** fields/funcs to the allocator struct

This means that user libraries may set 1.22.0 as their minimum required NumPy version, but still update their handler structs to match the latest NumPy release.

> Hmm, means we need to do `if/else` chains.

`if/else` chains would be also required in this solution, if we want newer NumPy versions to accept older handler allocator structs. Newer NumPy versions still need to know if/which e.g. functions are missing from a handler, and apply defaults. Of course there won't be string comparisons :).

> And I don't really expect more versions than maybe `2` and `3` ;).

I agree.

TLDR; If we think that future versions (most probably) are going to be extensions to the current struct (e.g. extra functions), a `version` field will simplify a lot its maintenance (both in user and NumPy side). Capsule names would be useful in a more complex scenario.

**Comment by jakirkham** [comment]
Matti added PR ( https://github.com/numpy/numpy/pull/20343 ) to include versioning. Would be great if others here took a look 🙂

**Comment by charris** [comment]
Is it OK to push this off? We now have versioning, and I don't think all the tasks here will be finished in time.

**Comment by seberg** [comment]
Yeah, should be good now.

**Comment by mattip** [comment]
Closing, I think we have resolved enough of these problems, please reopen or open a new issue.

**Comment by mattip** [comment]
The documentation added is https://numpy.org/devdocs/reference/c-api/data_memory.html, and there is also the NEP https://numpy.org/neps/nep-0049.html
